lib/python/picongpu/picmi/diagnostics/binning.py

```diff
@@ -0,0 +1,191 @@
+"""
+This file is part of PIConGPU.
+Copyright 2025 PIConGPU contributors
+Authors: Julian Lenz
+License: GPLv3+
+"""
+
+import sympy
+from ...pypicongpu.output.binning import (
+    Binning as PyPIConGPUBinning,
+    BinningFunctor as PyPIConGPUBinningFunctor,
+    BinningAxis as PyPIConGPUBinningAxis,
+    BinSpec as PyPIConGPUBinSpec,
+)
+from ...pypicongpu.species.species import Species as PyPIConGPUSpecies
+
+from ..species import Species as PICMISpecies
+from typing import Callable, Any, Optional
+from .timestepspec import TimeStepSpec
+import typeguard
+
+_COORDINATE_SYSTEM = {
+    (
+        origin.lower(),
+        precision.lower(),
+        unit.lower(),
+    ): tuple(sympy.Symbol(f"{c}_{precision.lower()}_{unit.lower()}") for c in coords)
+    for (origin, coords) in (
+        ("TOTAL", ("xt", "yt", "zt")),
+        ("GLOBAL", ("xg", "yg", "zg")),
+        ("LOCAL", ("xl", "yl", "zl")),
+        ("MOVING_WINDOW", ("xmw", "ymw", "zmw")),
+        ("LOCAL_WITH_GUARDS", ("xlg", "ylg", "zlg")),
+    )
+    for precision in ("CELL", "SUB_CELL")
+    for unit in ("CELL", "PIC", "SI")
+}
+
+
+class BinningParticle:
+    def __init__(self):
+        self.used_attributes = {}
+
+    def get_attribute_map(self):
+        return self.used_attributes
+
+    def get(self, attribute, **kwargs):
+        if attribute == "position":
+            origin = kwargs.get("origin", "total")
+            precision = kwargs.get("precision", "cell")
+            unit = kwargs.get("unit", "cell")
+            symbols = _COORDINATE_SYSTEM[(origin, precision, unit)]
+            self.used_attributes |= {symbols: ("position", origin, precision, unit)}
+
+        elif attribute == "momentum":
+            symbols = sympy.symbols("px,py,pz")
+            self.used_attributes |= {symbols: "momentum"}
+
+        elif attribute in ["gamma", "kinetic energy", "velocity"]:
+            # This relies on python dictionaries having a stable ordering.
+            # We first add mass and momentum
+            # and later use their symbols inside of the same preamble.
+            self.get("mass")
+            self.get("momentum")
+            if attribute == "gamma":
+                symbols = sympy.Symbol("gamma")
+            if attribute == "kinetic energy":
```

Comment:
```suggestion
            elif attribute == "kinetic energy":
```
Minor thing, but logically `elif` seems more appropriate. Same thing also happens below with velocity.